### DISCUSSION TITLE
How to detect if a Menu is Active and/or included

### DISCUSSION THREAD

**Discussion by Vodhin** [e107inc/e107#5256] [Q&A Core] [answered]
My Estate plugin has an optional Menu (estate_menu.php) file that is basically the same as a section of a listing's view page. I would like to detect if that menu has been activated and included when loading the view page: If it is active then I want to skip that bit of code. I can do this using JavaScript to remove one or the other, but I'd rather not have to do that.

**Comment by tgtje** [comment]
Maybe not what you like to see, but this ([old one](https://github.com/Jimako-e107-plugins/e107-old-plugins/blob/master/userjournals_menu/userjournals_menu.php)) part of repo from @Jimmi08 ; maybe of use? It's the only part i found used in e107 (afaik), that comes close i guess.

**Comment by Jimmi08** [comment]
https://github.com/e107inc/e107/issues/1871

e107::getMenu()->isLoaded('yourmenu') should work
 

Personally, I would put the code into class and check if class is already loaded or no. 

it is similar like the latest news menu on news plugin pages 
or vstore  menus on vstore pages   

you can check in menu e_CURRENTPLUGIN (or something like that), e_PAGE,  e_ROUTE...  

Be aware that menu can be loaded before or after main page (left on right side).  

I had/have maybe 10 issues related to this topic.   If I had needed this now, I would use query to database for actual THEME_LAYOUT if menu is active for this layout.

**Comment by rica-carv** [comment, marked as answer]
> e107::getMenu()->isLoaded('yourmenu') should work

@Jimmi08 Personally i use this method and never failed me. I thinks it's faster than anything else...

**Comment by Vodhin** [comment]
Thanks for the suggestions. @rica-carv , @Jimmi08 , and @tgtje Like @Jimmi08  mentioned, there doesn't seem to be an easy way to do what I need, which is NOT load something in the main body IF it will appear in a menu that renders AFTER the main body. 

I have tried a number of tricks but it looks like the best thing to do is make the menu.php simply return a target div and then use JavaScript to move it there if the menu is rendered.

**Comment by rica-carv** [comment]
@Vodhin I've used my method in the Voux theme and the main theme, and it works, with menu in area 2. What theme are you using?

**Reply by Vodhin** [reply]
I intend this to work for any theme, and have tested with bootstrap 3, 5, Voux (Which I've modified a bit to make the top static and fade out as you scroll down: see a live version of the plugin on [my client's website ](https://www.sandpiperhome.org/)). 

Anyway `e107::getMenu()->isLoaded('yourmenu') ` apparently needs to be used after `require_once(HEADERF); ` which meant I had to do a bit of code rearranging (I had been using the shortcode for the plugin to generate some inline "above the fold" CSS and JavaScript for the map section). 

Regardless, it took more than just getMenu() to do what I need, which is to see if the menu is loaded, prevent that menu from rendering the Agent Information in a Full Width layout (because I want it in the main body), prevent the Agent Information from being added to the Main page if it is in the Menu, and then prevent duplicates of the Agent Information because there is a JavaScript powered Contact form included (apparently it is possible assign the same menu to multiple areas in a theme).

So... what does work is pretty much this:

 In the `listing.php` file:
```

      $sc = e107::getScBatch('estate',true);
      $sc->setVars($PROPDTA[0]); // MUST BE BEFORE HEADERF so Menus will work
      require_once(HEADERF);
      $tmpl = e107::getTemplate('estate'); // MUST BE AFTER HEADERF so template gets THEME_LAYOUT and other stuff
      $text .= $tp->parseTemplate($tmpl['view']['top'], false, $sc);
      $text .= $tp->parseTemplate($tmpl['view']['sum'], false, $sc);
      etcetera...
```

And in the  _template_ file :
```
$EST_MENU1 = e107::getMenu()->isLoaded('estate');

$ESTATE_TEMPLATE['menu']['seller']= '{AGENT_CARD} and other stuff in the menu';


$ESTATE_TEMPLATE['view']['sum'] = 'Normal stuff here...'
if($EST_MENU1 == 0 || strpos(strtolower(THEME_LAYOUT),'full') !== false){ //do render here is Full Width 
  $ESTATE_TEMPLATE['view']['sum'] .= ' {AGENT_CARD} and Sidebar stuff that would be in the menu';
  define("ESTAGENTRENDERED",1);
  }
$ESTATE_TEMPLATE['view']['sum'] .= ' Rest of normal stuff...';
```

and in the estate_menu.php file:  
```
if($qs[0] == 'view'){
  if(strpos(strtolower(THEME_LAYOUT),'full') !== false){return '';} // do not render here if Full Width
  
  $thd = $tp->parseTemplate($tmpl['menu']['head'], false, $sc);
  $txt = $tp->parseTemplate($tmpl['menu']['seller'], false, $sc);
  $ns->tablerender($thd, $txt, 'estSideMenu1a');
  unset($thd,$txt);
  define("ESTAGENTRENDERED",1);
  
  $txt = $tp->parseTemplate($tmpl['menu']['events'], false, $sc);
  $ns->tablerender(EST_GEN_EVENTS, $txt, 'estSideMenu1b');
  unset($txt);
  }
```

and finally, in the shortcode.php file:
```
function sc_agent_card($parm){
   if(defined("ESTAGENTRENDERED")){return '<div class="estAgentCardDupe"></div>';}
   //agent information...
   }

```
